keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -393,6 +367,60 @@ of the following scenarios:
 2. A new Node results in a Service that is able to achieve an endpoint
    distribution below 20% for the first time.
 
+### Proportional CPU Heuristic
+
+#### Assumptions
+
+- Incoming traffic is proportional to the number of allocatable CPU cores in a
+  zone. Although this is an imperfect metric, it is the best available way of
+  predicting how much traffic will be received in a zone. If we are unable to
+  derive the number of allocatable cores in a zone we will fall back to the
+  number of nodes in that zone.
+- Service capacity is proportional to the number of endpoints in a zone. This
+  assumes that each endpoint has equivalent capacity. Although this is not
+  always true, it usually is. We can explore ways to deal with variable capacity
+  endpoints in the future.
+#### Example
+
+zone-a: 20 CPU cores
+zone-b: 16 CPU cores
+zone-c: 14 CPU cores
+
+In this scenario, the following proportion of endpoints would be allocated for
+each Service:
+
+zone-a: 40%
+zone-b: 32%
+zone-c: 28%
+
+When allocating endpoints to meet this distribution, keeping endpoints in the
+same zone will be prioritized. When same-zone endpoints are exhausted, endpoints
+will be taken from zones that have excess capacity.
+
+### Proportional by Zone Heuristic
+
+#### Assumptions
+
+- Incoming traffic is the same for all zones.
+- Each endpoint has equivalent capacity.
+
+This avoids the problem of blackholing traffic in case some zone run out of endpoints,
+but is still an imperfect method to achieve perfect traffic distribution.
+
+#### Example
+
+zone-a: 2 endpoints
+zone-b: 1 endpoint
+zone-c: 3 endpoints
```

Comment:
What happens if I have 4 Endpoints distributed such as:

```
zone-a: 1 endpoints
zone-b: 1 endpoint
zone-c: 2 endpoints
```
?

Will the hints be removed again as it is not possible to achieve a balanced distribution of endpoints across the zones?

We are looking into a deterministic way to have the hints set. We already use topologySpreadConstraints and nodeAffinity to spread the endpoints evenly across the zones. The endpoints will be spread evenly when the `number of endpoints % number of zones = 0`. But when the `number of endpoints % number of zones != 0` then there will be some zones with more replicas than the other. Or, in case of Deployment/StatefuleSet rollout, then the even distribution can be also violated.
Currently we have a mutating webhook that mutates the EndpointSlice's hints such as: it always sets the endpoint's hints to the endpoint's zone.
Shouldn't we consider adding this strategy as well? Or can you elaborate the newly added heuristic whether it is close to it? Because I am confused and cannot judge about it.
Assumptions for this strategy:
- You already have guaranteed appr. balanced distribution of endpoints across zones via scheduling means (topologySpreadConstraint, nodeAffinity)
- Each endpoint has equivalent capacity.

What the strategy does is to maintain the endpoint's zone as hint.

Reply:
> Currently we have a mutating webhook that mutates the EndpointSlice's hints such as: it always sets the endpoint's hints to the endpoint's zone.

sorry for the late response, I was suppose to close this PR to send a new one with the simple heuristic you mention, to copy the zone to the hint for the reasons that Rob mentions https://github.com/kubernetes/enhancements/pull/4003#issuecomment-1548807315 (specially 2.) but then we need to find a good solution to be completely sure we can handle Tim's comments https://github.com/kubernetes/enhancements/pull/4003#issuecomment-1549969610

@ialidzhikov one curiosity, how is copying over directly the zone to hints working for you? is that something that is completely solving your problems? 
cc: @robscott

Reply:
> @ialidzhikov one curiosity, how is copying over directly the zone to hints working for you? is that something that is completely solving your problems?

Some details for our setup are revealed in https://github.com/kubernetes/kubernetes/issues/110714#issuecomment-1164267496. For example, one of our use-case is to use topology-aware routing for the communication between the kube-apiserver and etcd. We run etcd in 3 zones, 1 replica per zone. The kube-apiserver replicas are also spread in similar way but the kube-apiserver replicas may vary between 3/4 replicas. The idea is that each kube-apiserver talks to the etcd Pod in its zone.
We also use topology-aware routing for the webhook communication. When kube-apiserver needs to talk to webhook deployed in the same cluster that the kube-apiserver runs. So, again, kube-apiserver talks to the webhook endpoint that is located in its zone (if there is such, of course).

Reply:
Just to make sure that I understand the `PreferZone` heuristic. It will always maintain the endpoint's zone as its hint, right?